src/library_async.js

```diff
@@ -410,6 +410,101 @@ mergeInto(LibraryManager.library, {
     });
   },
 
+  /*
+   * Layout of an EMTERPRETIFY_ASYNC coroutine structure:
+   *
+   *  0 callee's EMTSTACKTOP
+   *  4 callee's EMTSTACKTOP from Module['asm']
+   *  8 callee's EMT_STACK_MAX
+   * 12 my EMTSTACKTOP
+   * 16 my EMTSTACKTOP from Module['asm']
+   * 20 my EMT_STACK_MAX
+   * 24 coroutine function (0 if already started)
+   * 28 coroutine arg
+   * 32 my stack:
+   *    ...
+   */
+  emscripten_coroutine_create__sig: 'iii',
+  emscripten_coroutine_create__asm: true,
+  emscripten_coroutine_create__deps: ['malloc'],
+  emscripten_coroutine_create: function(f, arg, stack_size) {
+    f = f|0;
+    arg = arg|0;
+    stack_size = stack_size|0;
+    var coroutine = 0;
+
+    if ((stack_size|0) <= 0) stack_size = 4096;
+
+    coroutine = _malloc(stack_size + 32)|0;
+    {{{ makeSetValueAsm('coroutine', 12, '(coroutine+32)', 'i32') }}};
+    {{{ makeSetValueAsm('coroutine', 16, '(coroutine+32)', 'i32') }}};
+    {{{ makeSetValueAsm('coroutine', 20, '(coroutine+32+stack_size)', 'i32') }}};
+    {{{ makeSetValueAsm('coroutine', 24, 'f', 'i32') }}};
+    {{{ makeSetValueAsm('coroutine', 28, 'arg', 'i32') }}};
+    return coroutine|0;
+  },
+
+  emscripten_coroutine_next__sig: 'ii',
+  emscripten_coroutine_next__deps: ['$EmterpreterAsync', 'free'],
+  emscripten_coroutine_next: function(coroutine) {
+    // this is a rewritten emscripten_coroutine_next function from ASYNCIFY
+    coroutine = coroutine|0;
+    var temp = 0, func = 0, funcArg = 0, coroutine_not_finished = 0;
+
+    // switch context
+    {{{ makeSetValueAsm('coroutine', 0, 'EMTSTACKTOP', 'i32') }}};
+    temp = Module['asm'].emtStackSave();
```

Comment:
these are unsafe - using `asm` is not supported. instead, those should be exported, and usable as `Module['emtStackSave']()` etc.

Reply:
(In other words, `Module['asm']` is an internal detail, that should not be relied on, as little as possible.)

Reply:
I have rewritten calls in my PR, but `EmterpreterAsync.handle()` does already contain such calls -- should they be converted too? And what about allocation size in the coroutine creation function from ASYNCIFY and function signatures?

Reply:
Yes, if there are other things like that, we should convert them as well.

Not sure what you mean about allocation size? If you mean that asyncify also does that, then yes, the same principle applies there, we should use things on `Module`, avoiding `Module['asm']`.

Reply:
I mean shouldn't we [here](https://github.com/atrosinenko/emscripten/blob/b8e4f8c3cfe1e9b043bd4213ad9a960eed4fcd7e/src/library_async.js#L134) use `coroutine = _malloc((stack_size + 32) | 0)|0;` instead?

Reply:
I see, thanks. Yes, you're right, the allocation should add the stack plus the extra space we need.